src/librustc_typeck/check/closure.rs

```diff
@@ -269,4 +273,408 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+///
+/// The expectation matcher is always applied to types, E and S,
+/// where E is an input/output type from the expected signature,
+/// and S is the corresponding type that the user supplied. Note that
+/// the user may not provide types for all inputs/outputs -- e.g.,
+/// in a call like this:
+///
+///     foo(|x, y: u32| -> i32 { .. })
+///
+/// the user supplied values for `y` and the return type, but not
+/// `x`. In that case, if ESig is `(T1, T2) -> T3`, the
+/// `ExpectationMatcher` will be invoked on the pairs (T2, u32) and
+/// (T3, i32) (note that T1 does not participate).
+///
+/// For each pair (E0, S0) of types, we will relate the types
+/// structurally, producing a new type R that will be used for the
+/// type in the final signature. One subtle point: both types may
+/// contain late-bound regions with depth 1: these late-bound regions
+/// will be bound by the closure signature.
+///
+/// If all the type pairs in the signatures are relatable, we are
+/// attempting to ensure two key invariants regarding the results:
+///
+/// - First, each type R only contains **bound regions that
+///   originated from the corresponding expected type E**.
+///   - That is, if the user wrote `|x: &str, y|`, the resulting bound
+///     region name in `x` (`Anon(0)`) would not be combined with some
+///     bound region name that appears in the expected type for `y`. The bound region
+///     names from the expected/supplied types effectively inherit distinct
+///     namespaces, and should never be mixed.
+/// - Second, any bound region appears that appears in some expected type E,
+///   also appears in the corresponding spot in the resulting type R.
+///   - In particular, imagine that we have an expected signature
+///     `(&str) -> &str`, but the user wrote `|x: &'a str| ...`. In
+///     the supplied type, the region `'a` is free.
+///     - We could choose to use the final signature `(&'a str) ->
+///       &str` without violating the first invariant, since we did
+///       "intermix" any bound region names from the supplied types
+///       with those of the expected type.
+///     - However, if we did so, we have lost the constraint expressed
+///       in the expected signature, that the return type must relate
+///       to the input.  In fact, this resulting signature is
+///       inconsistent, since a bound region cannot only appear in the
+///       input (this was #38714).
+///
+/// To produce the type R, we relate the types recursively, which
+/// requires that the types generally match. If that is not true, an
+/// error results, which will result in the expected signature being
+/// ignored. The places where the types do not have to match, and the
+/// resulting type in those scenarios, are as follows:
+///
+/// - If the expected type E is an (unbound) inference variable,
+///   and S is not an inference variable, then the resulting type will be S,
+///   so long as S does not contain anonymous regions.
```

Comment:
And if `S` does contain anonymous regions, what happens (this needs to be specified).

Reply:
I will try to make this clearer. The intention was that, if none of the conditions in this list apply, then the "supplied types" will be used "as is" as the resulting closure type. Typically, this will then result in an error unless the "expected type" is precisely equal (because of invariance).

So, in this particular case, if S contains anonymous regions, then the supplied types (for all parameters) are used "as is".

Reply:
(At least that's how I remember it. It's been a while, so I'll double check.)